plugin/src/main/java/org/opensearch/ml/action/stats/MLStatsNodesTransportAction.java

```diff
@@ -126,22 +146,72 @@ MLStatsNodeResponse createMLStatsNodeResponse(MLStatsNodesRequest mlStatsNodesRe
             }
         }
 
+        CountDownLatch latch = new CountDownLatch(mlStats.getAllModels().length);
         Map<String, MLModelStats> modelStats = new HashMap<>();
         // return model level stats
         if (mlStatsInput.includeModelStats()) {
             for (String modelId : mlStats.getAllModels()) {
-                if (mlStatsInput.retrieveStatsForModel(modelId)) {
-                    Map<ActionName, MLActionStats> actionStatsMap = new HashMap<>();
-                    for (Map.Entry<ActionName, MLActionStats> entry : mlStats.getModelStats(modelId).entrySet()) {
-                        if (mlStatsInput.retrieveStatsForAction(entry.getKey())) {
-                            actionStatsMap.put(entry.getKey(), entry.getValue());
+                // Add action listener to retrieve model details
+                validateAccess(modelId, ActionListener.wrap(hasPermissionToShowStat -> {
+                    if (hasPermissionToShowStat) {
+                        if (mlStatsInput.retrieveStatsForModel(modelId)) {
+                            Map<ActionName, MLActionStats> actionStatsMap = new HashMap<>();
+                            for (Map.Entry<ActionName, MLActionStats> entry : mlStats.getModelStats(modelId).entrySet()) {
+                                if (mlStatsInput.retrieveStatsForAction(entry.getKey())) {
+                                    actionStatsMap.put(entry.getKey(), entry.getValue());
+                                }
+                            }
+                            modelStats.put(modelId, new MLModelStats(actionStatsMap));
                         }
                     }
-                    modelStats.put(modelId, new MLModelStats(actionStatsMap));
-                }
+                    // Count down the latch after each asynchronous call completes
+                    latch.countDown();
+                }, e -> {
+                    // Handle failure case here
+                    // For example, log the error
+                    log.error("Failed to retrieve model details for model ID: " + modelId, e);
+                    // Count down the latch even in case of failure to ensure proper synchronization
+                    latch.countDown();
+                }));
+            }
+            // Wait for all asynchronous calls to complete
+            try {
+                latch.await();
+            } catch (InterruptedException e) {
+                // Handle interruption if necessary
+                Thread.currentThread().interrupt();
```

Comment:
Why not use await(long timeout, TimeUnit unit) to avoid infinite waiting loop?

Reply:
added default timeout for 2 minutes.

Reply:
In the case of timeout, the results will be the partial stats for all models? e.g. some of model stats are not included in the results?

Reply:
Yeah that's correct. But I don't think it will take 2 minutes to get all the models. What do you think?

Reply:
Is it trivial to add a flag in the response to show if the stats are complete in summarizing the MLStatsNodesResponse from all nodes? Return partial is fine, but it would be better to indicate that in the response.

Reply:
It don't think it's trivial and also I'm not sure if it's necessary though. If we think 2 minutes isn't enough threshold time, we can increase it to 5 minutes. But in my experience, It didn't take too long to get 1000 models in stat api response.

Reply:
Your assumption may be valid as of now, but not necessarily true if more stats are added later on in the code. Also if this stats API is running on a schedule that is less than 2 minutes, is it going to trigger problems and delay the response?

Reply:
>Also if this stats API is running on a schedule that is less than 2 minutes

Can you explain a bit more here?

>but not necessarily true if more stats are added later on in the code

how more stats are we expecting?

Reply:
For example, at some point we will need to query the stats periodically and send the data as metrics to CW for dashboard monitoring, etc. Or at some time, the stats API is called at a large TPS. The waiting in the thread will be a potential problem? We cannot just limit this API to be a very simple use case right?